File: Python/algorithmic_heights/22_bf.py
# Bellman-Ford Algorithm
import sys

from collections import defaultdict, namedtuple
from math import inf, isinf, isfinite

Edge = namedtuple('Edge', ['node_from', 'node_to', 'weight'])


class DirectedGraph:
    def __init__(self, nr_nodes, edge_list):
        self._nr_nodes = nr_nodes
        self._adjacency_list = self._build_adjacency_list(edge_list)

    def __iter__(self):
        yield from self._adjacency_list.items()

    @staticmethod
    def _build_adjacency_list(edge_list):
        adjacency_list = defaultdict(list)
        for edge in edge_list:
            adjacency_list[edge.node_from].append(edge)
        return dict(adjacency_list)

    @property
    def nr_nodes(self):
        return self._nr_nodes


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        a, b, w = convert_to_intlist(next(reader))
        edges.append(Edge(node_from=a, node_to=b, weight=w))
    return edges


def update_distances(graph, distances):
    for node, neighbours in graph:
        dist_node = distances[node-1]
        if isfinite(dist_node):
            for _, neighbour, weight in neighbours:
                distance_through_node = dist_node + weight
                if distances[neighbour-1] > distance_through_node:
                    distances[neighbour-1] = distance_through_node


def bellman_ford_algorithm(graph, node):
    n = graph.nr_nodes
    distances = [inf] * n
    distances[node-1] = 0
    for _ in range(n):
        update_distances(graph, distances)
    return distances


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_edges(reader, nr_edges)
    graph = DirectedGraph(nr_nodes, edge_list)
    result = bellman_ford_algorithm(graph, node=1)
    print(' '.join(map(lambda x: 'x' if isinf(x) else str(x), result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/22_bf.hs
import Control.Monad (replicateM)
import Data.IntMap (IntMap)
import Data.IntSet (IntSet)
import qualified Data.IntMap as M
import qualified Data.IntSet as S

data Edge = Edge { nodeFrom :: Int, nodeTo :: Int, weight :: Int }
data Graph = Graph { nrNodes :: Int, adjacencyList :: IntMap [Edge] }

data Distance = Dist Int | Infinity deriving Eq

instance Ord Distance where
    Infinity <= Infinity = True
    Dist d1 <= Dist d2 = d1 <= d2
    Infinity <= Dist _ = False
    Dist _ <= Infinity = True


instance Show Distance where
    show (Dist d) = show d
    show Infinity = "x"


addDist :: Distance -> Distance -> Distance
addDist (Dist d1) (Dist d2) = Dist (d1 + d2)
addDist _ _ = Infinity


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readInputLine :: String -> (Int, Int)
readInputLine line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


readEdges :: Int -> IO [Edge]
readEdges nrEdges = do
    let parse line =
            case convertToIntList line of
                [a, b, w] -> Edge a b w
                _ -> error "Malformed edge input."
    edges <- replicateM nrEdges getLine
    return $ map parse edges


createGraph :: Int -> [Edge] -> Graph
createGraph nrNodes edges =
    let addEdge adjacencyList edge =
            let update Nothing = Just [edge]
                update (Just ns) = Just (edge : ns)
            in M.alter update (nodeFrom edge) adjacencyList
        adjacencyList = foldl addEdge M.empty edges
    in Graph { nrNodes = nrNodes, adjacencyList = adjacencyList }


updateDistances :: Graph -> IntMap Distance -> IntMap Distance
updateDistances graph distances =
    let updateSingleDistance dist currentDistances (Edge _ neighbor weight) =
            let distThroughNode = addDist dist (Dist weight)
                neighborDist = M.findWithDefault Infinity neighbor currentDistances
            in if neighborDist > distThroughNode then M.insert neighbor distThroughNode currentDistances
            else currentDistances
        update node neighbors currentDistances =
            case M.lookup node currentDistances of
                Nothing -> currentDistances
                Just distNode -> foldl (updateSingleDistance distNode) currentDistances neighbors
    in M.foldrWithKey update distances (adjacencyList graph)


bellmanFord :: Graph -> Int -> [Distance]
bellmanFord graph @ (Graph nrNodes adjacencyList) sourceNode =
    let distances = foldl (\acc _ -> updateDistances graph acc) (M.singleton sourceNode (Dist 0)) [1..nrNodes]
    in map (\node -> M.findWithDefault Infinity node distances) [1..nrNodes]


main :: IO ()
main = do
    (nrNodes, nrEdges) <- readInputLine <$> getLine
    edges <- readEdges nrEdges
    let graph = createGraph nrNodes edges
    let result = bellmanFord graph 1
    putStrLn $ unwords $ map show result
